File: 250319/트로미노/tromino.cpp
#include <bits/stdc++.h>
using namespace std;

#define MAX_NM 200
#define DIR_NUM 4

int dx[DIR_NUM] = {-1, 0, 1, 0}, dy[DIR_NUM] = {0, 1, 0, -1};
int n, m;
int grid[MAX_NM][MAX_NM];

int ans = 0;

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<m;
}

void Sum1(int x, int y) {


    for(int d=0; d<DIR_NUM; d++) {
        int sum = grid[x][y];

        int nx = x+dx[d], ny = y+dy[d];
        int nx2 = x+dx[(d+1)%DIR_NUM], ny2 = y+dy[(d+1)%DIR_NUM];

        if (InRange(nx, ny)) {
            sum += grid[nx][ny];
        }

        if (InRange(nx2, ny2)) {
            sum += grid[nx2][ny2];
        }

        ans = max(ans, sum);

    }
}

void Sum2(int x, int y) {
    for(int d=0; d<2; d++) {
        int sum = grid[x][y];

        int nx = x+dx[d], ny = y+dy[d];
        int nx2 = x+dx[(d+2)%DIR_NUM], ny2 = y+dy[(d+2)%DIR_NUM];

        if (InRange(nx, ny)) {
            sum += grid[nx][ny];
        }

        if (InRange(nx2, ny2)) {
            sum += grid[nx2][ny2];
        }

        ans = max(ans, sum);
    }
}

int main() {
    cin >> n>>m;
    
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];
        }
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            Sum1(i, j);
            Sum2(i, j);
        }
    }

    cout << ans;
    return 0;
}

File: 250922/트로미노/tromino.py
DIR_NUM = 4

n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

dx, dy = [-1, 0, 1, 0], [0, 1, 0, -1]

ans = 0

def in_range(x, y):
    return 0<=x<n and 0<=y<m

def sum1(x, y):
    result = 0

    for d in range(DIR_NUM):
        tmp = grid[x][y]

        nx1, ny1 = x+dx[d], y+dy[d]
        nx2, ny2 = x+dx[(d+1)%DIR_NUM], y+dy[(d+1)%DIR_NUM]

        if in_range(nx1, ny1) and in_range(nx2, ny2):
            result = max(result, tmp + grid[nx1][ny1] + grid[nx2][ny2])
    
    return result


def sum2(x, y):
    result = 0

    for d in range(2):
        tmp = grid[x][y]

        nx1, ny1 = x+dx[d], y+dy[d]
        nx2, ny2 = x+dx[(d+2)%DIR_NUM], y+dy[(d+2)%DIR_NUM]

        if in_range(nx1, ny1) and in_range(nx2, ny2):
            result = max(result, tmp + grid[nx1][ny1] + grid[nx2][ny2])
    
    return result


for i in range(n):
    for j in range(m):
        tmp = max(0, max(sum1(i, j), sum2(i, j)))

        ans = max(ans, tmp)

print(ans)

